File: src/main/java/me/nasukhov/TukitaLearner/bot/task/AskQuestion.java
package me.nasukhov.TukitaLearner.bot.task;

import me.nasukhov.TukitaLearner.bot.bridge.IOResolver;
import me.nasukhov.TukitaLearner.bot.io.Channel;
import me.nasukhov.TukitaLearner.bot.io.Output;
import me.nasukhov.TukitaLearner.study.GroupQuestion;
import me.nasukhov.TukitaLearner.study.ProgressTracker;
import me.nasukhov.TukitaLearner.study.Time;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Component("askQuestionTask")
public class AskQuestion implements TaskRunner {
    private final ProgressTracker progressTracker;
    private final IOResolver ioResolver;

    public AskQuestion(ProgressTracker progressTracker, IOResolver ioResolver) {
        this.progressTracker = progressTracker;
        this.ioResolver = ioResolver;
    }

    @Override
    public String subscribesFor() {
        return "ask_question";
    }

    @Override
    public void runTask(Task task) {
        if (!subscribesFor().equals(task.getName())) {
            throw new RuntimeException("Runner does not know how to execute the given task");
        }

        // We don't want it to work at night
        if (Time.isOffHours()) {
            return;
        }

        ask(task.getChannel());
    }

    public void ask(Channel channel) {
        Optional<GroupQuestion> result = progressTracker.createRandomForChannel(channel);
        if (result.isEmpty()) {
            return;
        }

        GroupQuestion newQuestion = result.get();

        Map<String, String> replies = new HashMap<>();
        int optionNum = 0;
        for (String replyVariant : newQuestion.listVariants()) {
            replies.put(replyVariant, String.format("qh answer %s %d", newQuestion.getId().toString(), ++optionNum));
        }

        Output output = ioResolver.resolveFor(channel);
        output.promptChoice(newQuestion.getText(), replies);
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/task/AskQuestion.kt
package me.nasukhov.tukitalearner.bot.task

import me.nasukhov.tukitalearner.bot.bridge.IOResolver
import me.nasukhov.tukitalearner.bot.io.Channel
import me.nasukhov.tukitalearner.study.ProgressTracker
import me.nasukhov.tukitalearner.study.Time.isOffHours
import org.springframework.stereotype.Component

@Component("askQuestionTask")
class AskQuestion(
    private val progressTracker: ProgressTracker,
    private val ioResolver: IOResolver,
) : TaskRunner {
    override fun subscribesFor(): String = "ask_question"

    override fun runTask(task: Task) {
        check(subscribesFor() == task.name) { "Runner does not know how to execute the given task" }

        // We don't want it to work at night
        if (isOffHours) {
            return
        }

        ask(task.channel)
    }

    fun ask(channel: Channel) {
        val result = progressTracker.createRandomForChannel(channel)
        if (result.isEmpty) {
            return
        }

        val newQuestion = result.get()

        val replies = HashMap<String, String>()
        var optionNum = 0
        for (replyVariant in newQuestion.listVariants()) {
            replies.put(replyVariant, "qh answer $newQuestion.id ${++optionNum}")
        }

        val output = ioResolver.resolveFor(channel)
        output.promptChoice(newQuestion.text, replies)
    }
}
